### ISSUE TITLE
multipart/form-data support

### ISSUE THREAD

**Issue by joshheinrichs** [PolymerElements/iron-form#54]
I'm trying to set the content-type to `multipart/form-data` so I can upload images to a Go server, but I'm getting the following error when trying to parse the form on my server:

```
no multipart boundary param in Content-Type
```

Is `multipart/form-data` not supported, or am I missing something? Here's a reduced example which produces the issue:

```HTML
<dom-module id="file-upload">
	<template>
		<form is="iron-form" id="form" content-type="multipart/form-data" action="http://localhost:9090/upload" method="post">	
			<input name="field-1" type="file" accept="image/*" required><br>
			<button on-tap="submit">submit</button>
		</form>
	</template>
</dom-module>
<script>
	Polymer({
		is: "file-upload",
		submit: function() {
			this.$.form.submit();
		}
	});
</script>
```

I should note that I can upload images using a regular form, so I know it isn't a problem with my server.

**Comment by vaadin-marcus** [comment]
I was struggling with the same for a long time today. Eventually, I had to give up on iron-form for submitting, instead doing a manual XMLHTTPRequest. Basically:
```
var form = document.getElementById('form-id');
var formData = new FormData(form);
var xhr = new XMLHttpRequest();
xhr.open('POST', '/upload/path', true);
xhr.send(formData);
```

I think the most elegant solution would be that the serialize function in iron-form would take advantage of `FormData` and set that as the body of the request. This will at the same time set the correct content type. 

I'm open to contributing, but I'm not really sure how the process works - what tests and other requirements there are.

**Comment by notwaldorf** [comment]
`iron-form` does not support multipart form data at the moment, but I am adding a presubmit event (https://github.com/PolymerElements/iron-form/issues/26), so that you can change the parameters of the request manually, before it's being sent to the server.

**Comment by eximius313** [comment]
+1

**Comment by notwaldorf** [comment]
This landed in #67

**Comment by eximius313** [comment]
@notwaldorf but it should work automatically or I need to add some presubmit code?

**Comment by notwaldorf** [comment]
@eximius313 You need to add presubmit code. I don't think `iron-ajax` supports multipart/form-data out of the box

**Comment by eximius313** [comment]
so shouldn't this issue be reopened until it starts to support it?

**Comment by ahedderich** [comment]
@eximius313 what you are looking for is more an issue with iron-ajax.
iron-form "only" gathers all information and submits it via iron-ajax. The iron-ajax element can't handle file type parameters.
I recently started building an multi purpose upload box that works with iron-form.
By using the pre-submit hook to change the iron-ajax request body to an FormData object, files are handled correctly.
```
document.querySelector('form').addEventListener('iron-form-presubmit',function(e){
			var data = e.target.request.body;
			var formdata = new FormData();
			for (var key in data) {
				if (data.hasOwnProperty(key)) {
					var value = data[key];
					if(Object.prototype.toString.call( value ) === '[object Array]'){
						for (var key2 in value) {
							formdata.append(key + '[' + key2 + ']', value[key2]);
						}
					}else{
						formdata.append(key, value);
					}
				}
			}
			e.target.request.contentType = undefined;
			e.target.request.body = formdata;
		});
```
Note, that the request.contentType needs to be undefined since iron-ajax would override it otherwise. The FormData object takes care of the correct ContentType and the necessary content boundary.

**Comment by eximius313** [comment]
@ahedderich Thank you

**Comment by wesleyduff** [comment]
Finnaly ran across this. Been trying this for about hour now. Would be nice to implement this feature in iron-ajax. The use of file uploading is rare so I guess using rolling your own will be fine as well. Thanks everyone :)

**Comment by Alvarz** [comment]
Hello,  `iron-ajax` doesn't support `enctype="multipart/form-data"` yet, but you can handle it with javascript's `FormData();`. You can create a function which convert the files to formdata and then append it to the body of `iron-ajax`.

    Polymer({
    ......
    convertFileInputToFormData: function(){
      regexp = /^[^[\]]+/;
      var fileInput = $('input[type="file"]');
      var fileInputName = regexp.exec( fileInput.attr('name') );
      // make files available
      var data = new FormData();
      jQuery.each($(fileInput)[0].files, function(i, file) {
          data.append(fileInputName+'['+i+']', file);
      });
      return data;
    },
    .......

and then you can call it from your pre-submit function like this
   
    formPreSubmit: function(event){
      var form = document.getElementById('Form');
        // here you convert to formData calling our function
        var data = this.convertFileInputToFormData();

        form.request.method = "post";
       // set undefined to prevent default json content type
        form.request.contentType = undefined;
       // here you append your formData to the iron-ajax body
        form.request.body = data;
      form.request.url = "http://your.api/url"

      form.request.debounceDuration = 300;
    },

The  `form.request.contentType = undefined;` prevent to `iron-ajax` set content type as json by default.

on the php file you must get the data like that


    if (isset($_FILES['file']['name'][0])) {
    
    $file = $_FILES['file']['name'][0];
        $type = $_FILES['file']['type'][0];
    // here yor upload methods
    
    }

the `$_FILES['file']['name'][0]` its because they can handle an array of image.
   

I hope this will help you and sorry my terrible english

**Comment by dman777** [comment]
Is it possible to send both form fields and uploaded images in the same form post? I am following the above example  but the images are clobbering the form fields in the network post request.... so only the images are included and not the form fields that did exist before:

```javascript
       this.formData.addEventListener('iron-form-presubmit', ()=> {
          bar.append('foo', this.formImages[0]);
          this.formData.request.body = bar;
          this.formData.request.contentType = undefined;
        });
```

**Comment by dman777** [comment]
Please make iron-forms support  multipart/form-data. The pre-submit work around causes havoc because it overwrites existing form data when attaching images.

**Comment by HendrikThePendric** [comment]
Could someone explain why this is even influencing the behaviour of the iron-form when `allow-redirect="true"` is set?

**Comment by stefanmerkel** [comment]
+1

**Comment by faan11** [comment]
+1

**Comment by krismatusiak** [comment]
+1

**Comment by spencerwmiles** [comment]
Why is this issue closed? This should  have been a user story from day one... Terrible to have to realize this 3/4th the way through development.

**Comment by ameykshirsagar** [comment]
Hello everyone
I have written a small blog post on a working method to upload files properly
[https://arcoirislabs.medium.com/uploading-files-in-polymer-3-using-iron-form-e10ecf8509ea?sk=650511425706ac72df1ed29927bc7a8b](url)
Kindly check it out
